core/src/main/java/org/apache/accumulo/core/conf/Property.java

```diff
@@ -1069,6 +1075,11 @@ public enum Property {
           + "also consider configuring the `" + NoDeleteConstraint.class.getName() + "` "
           + "constraint.",
       "2.0.0"),
+  TABLE_ASSIGNMENT_GROUP("table.assignment.group", Constants.DEFAULT_RESOURCE_GROUP_NAME,
```

Comment:
it's also used in TabletGroupWatcher.getAssignmentsFromBalancer to print a warning if a tablet is assigned to a tserver outside of its resource group.

Reply:
> it's also used in TabletGroupWatcher.getAssignmentsFromBalancer to print a warning if a tablet is assigned to a tserver outside of its resource group.

I don't TGW should try to validate if a plugin is behaving properly.

Reply:
If we scope the config to plugin, it could allow TableLoadBalancer to have more complex configuration that is tightly coupled to it.  For example TableLoadBalancer could support the property `table.custom.assignment.group` or `table.custom.assignment.rangedGroups` where rangedGroups allows a user to set a json value that specifies something like the folllowing.

 *  For tablets in range `(<row1>,<row2>]` user tablet server group X
 *  For tablets in range `(<row3>,<row4>]` user tablet server group Y
 * For all other tablet use tablet server group Z

So one property or the other could be set, if both are set it throws an error.

Reply:
So, it wasn't just the warning in TGW.getAssignmentsFromBalancer. There is also this:

https://github.com/apache/accumulo/pull/3496/files#diff-dd97054ac43737177293a8f3a9231f05f0e4bc5eb0a3b042cad3fb8ce31b6caaR303-R325

Reply:
> So, it wasn't just the warning in TGW.getAssignmentsFromBalancer. There is also this:
>
> https://github.com/apache/accumulo/pull/3496/files#diff-dd97054ac43737177293a8f3a9231f05f0e4bc5eb0a3b042cad3fb8ce31b6caaR303-R325

For that case maybe we need to add a new method to the balancer so we can ask `hey balancer are you happy with the group this tablet is currently assigned to?`

Reply:
>  So, it wasn't just the warning in TGW.getAssignmentsFromBalancer. There is also this:
> 
> https://github.com/apache/accumulo/pull/3496/files#diff-dd97054ac43737177293a8f3a9231f05f0e4bc5eb0a3b042cad3fb8ce31b6caaR303-R325


Does this check also need to run in the tablet mgmt iterator?

Reply:
> For that case maybe we need to add a new method to the balancer so we can ask hey balancer are you happy with the group this tablet is currently assigned to?

Wondering the balancers existing `long balance(BalanceParameters params);` method can handle this case.  Can it efficiently detect tablets assigned to an undesired group and initiate migrations?

Reply:
I don't know, I think I put the TODO there to have a conversation about it. I'm going to take another look at this before I merge.

Reply:
Yeah it needs investigation, need to figure out how to defer to the balancer plugin for this and if that needs new methods in the balancer.

Reply:
I was thinking a new interface that TableLoadBalancer implements and HostRegexTableLoadBalancer does not. Manager could check and invoke the methods on it.

Reply:
> Does this check also need to run in the tablet mgmt iterator?

I don't think it can unless we pass the group information to the iterator. The group is not in the metadata anymore, only in ZK. The LiveTServerSet object in the Manager is the only thing that knows about the TServer groups

Reply:
If needed we could  pass the tserver group information to the tablet mgmt iterator, we already pass the live tserver set I think so this would be a similar amount of data.  We could call the balancer plugins in the tablet mgmt iterator fi needed also.  The compaction code is calling compaction plugins in CompactionJobGenerator inside tablet mgtm iterator to make decisions.

Reply:
It may be good to have an IT that does something like the following.

 * Configures table A to be assigned to tserver group Z
 * Verifies assignments are as expected
 * Changes config such that table A is assigned to tserver group X
 * Starts a tserver in group X
 * Waits for tablets to be assigned to group X

Reply:
I removed the TABLE_ASSIGNMENT_GROUP property from Property and kept it in TableLoadBalancer in the commits today. There is a lot in this thread. I need to re-read it and make sure I got to everything.

Reply:
I added a test that you suggested above. It uncovered some other issues with tables being deleted and how the TabletManagementIterator handles that. I think I have everything in this thread done.